radicle-surf/src/vcs/git/repo.rs

```diff
@@ -84,56 +230,50 @@ impl<'a> RepositoryRef<'a> {
-            .try_fold(vec![], |mut acc, reference| {
-                let branch = Branch::try_from(reference?)?;
-                acc.push(branch);
-                Ok(acc)
-            })
+    /// Returns an iterator of branches that match `pattern`.
+    pub fn branches(&self, pattern: &Glob<Branch>) -> Result<Branches, Error> {
+        let mut branches = Branches {
+            references: vec![],
+            current: 0,
+        };
+        for glob in pattern.globs().iter() {
+            let namespaced = self.namespaced_refname(glob)?;
+            let references = self.repo_ref.references_glob(&namespaced)?;
+            branches.references.push(references);
+        }
+        Ok(branches)
     }
 
-    /// List the tags within a repository, filtering out ones that do not parse
-    /// correctly.
-    ///
-    /// # Errors
-    ///
-    /// * [`Error::Git`]
-    pub fn list_tags(&self, scope: RefScope) -> Result<Vec<Tag>, Error> {
-        RefGlob::tag(scope)
-            .references(self)?
-            .iter()
-            .try_fold(vec![], |mut acc, reference| {
-                let tag = Tag::try_from(reference?)?;
-                acc.push(tag);
-                Ok(acc)
-            })
+    /// Returns an iterator of tags that match `pattern`.
+    pub fn tags(&self, pattern: &Glob<Tag>) -> Result<Tags, Error> {
+        let mut tags = Tags {
+            references: vec![],
+            current: 0,
+        };
+        for glob in pattern.globs().iter() {
+            let namespaced = self.namespaced_refname(glob)?;
+            let references = self.repo_ref.references_glob(&namespaced)?;
+            tags.references.push(references);
+        }
+        Ok(tags)
     }
 
-    /// List the namespaces within a repository, filtering out ones that do not
-    /// parse correctly.
-    ///
-    /// # Errors
-    ///
-    /// * [`Error::Git`]
-    pub fn list_namespaces(&self) -> Result<Vec<Namespace>, Error> {
-        let namespaces: Result<HashSet<Namespace>, Error> = RefGlob::Namespace
-            .references(self)?
-            .iter()
-            .try_fold(HashSet::new(), |mut acc, reference| {
-                let namespace = Namespace::try_from(reference?)?;
-                acc.insert(namespace);
-                Ok(acc)
-            });
-        Ok(namespaces?.into_iter().collect())
+    /// Returns the list of namespaces that match `pattern`.
+    pub fn namespaces(&self, pattern: &Glob<Namespace>) -> Result<Vec<Namespace>, Error> {
+        let mut set = HashSet::new();
+        for glob in pattern.globs.iter() {
+            let new_set = self
+                .repo_ref
+                .references_glob(glob.as_str())?
+                .map(|reference| {
+                    reference
+                        .map_err(Error::Git)
+                        .and_then(|r| Namespace::try_from(r).map_err(|_| Error::EmptyNamespace))
+                })
+                .collect::<Result<HashSet<Namespace>, Error>>()?;
```

Comment:
```suggestion
                .collect::<Result<BTreeSet<Namespace>, Error>>()?;
```

I tend to prefer `BTreeSet` because I think the hasher is a bit expensive, iirc

Reply:
Will update.